remove opam dependancy

RyanGibb opened the pull request:
if the `opam config subst` call fails, substitute the jobs variable with the number of physical cores in the machine

TheLortex reviewed (changes requested):
Looks good. The context here is that when used in a `nix` build, the `opam` command is not available (or available but with no switch bound to it), so we need a default configuration for the number of jobs used in the build.

RyanGibb commented on src/dune:
```diff
@@ -23,8 +23,10 @@
  (mode fallback)
  (action
-  (run opam config subst "build.sh")))
+  (run sh -c "opam consfig subst %{target} ||
```
Fixed, thank you!

hannesm commented:
I'm sorry to be picky, but `/proc/cpuinfo` is only available on GNU/Linux, not on any other Unix. If you want to find out how many CPUs are part of the hardware, please consult opam, or OCaml (e.g. https://github.com/ocaml/ocaml/pull/11309).

Or, in case of `opam` not being available, why not simply use `1`?

TheLortex commented:
Yes we figured the portability issue. Another solution is to use `sysctl` to obtain CPU settings.

GMP is quite long to build so parallel jobs are very useful, hence the choice to use more than one job when possible. The reason why `opam` is not available is that @RyanGibb is working on building unikernels using Nix.

RyanGibb commented:
Yes, I [tried](https://github.com/RyanGibb/hillingar/commit/eac2b06e9ff13d4b836f84252fef9c20fcbf7880) to build an opam switch in a nix derivation but there doesn't seem to be a nice way to install the compiler without network access.

> Or, in case of opam not being available, why not simply use 1?

I've just pushed a change that defaults to 1 if `/proc/cpuinfo` isn't available.

> Another solution is to use sysctl to obtain CPU settings.

It seems like the variable for CPU cores in sysctl differs by platform (`sysctl hw.ncpu` seems to work on BSD derivatives but not Linux).

This shouldn't be an issue outside though. Where opam is available outside nix derivations this will continue working as normal. On Linux the number of physical cores will be used, and otherwise it will fall back on 1.

dinosaure commented:
> The reason why opam is not available is that @RyanGibb is working on building unikernels using Nix.

It seems rather counter-intuitive to me when the `mirage` workflow requires `opam' in anyway.

hannesm commented:
I'm all in for removing the opam dependency from the build of `ocaml-gmp`. The background is that I also compile MirageOS unikernels without an `opam` binary. (see https://github.com/roburio/orb/commit/ed60f97cfe01228b8dae78a2be90b6f50ba14191 for the specific hack I had to implement just for ocaml-gmp).

Now, on the downside I'm against adding complexity (read: conditionals) to the build systems. So, if it is desired to use multiple CPU cores while building ocaml-gmp, why not resort to the opam package processor (https://github.com/haesbaert/ocaml-processor) that provides a function `cpu_count` and is available for all OCaml versions and platforms ocaml-gmp is supposed to work with.

hannesm commented:
> It seems rather counter-intuitive to me when the `mirage` workflow requires `opam` in anyway.

It actually does not really anymore -- I mean truly there's the opam-repository, and `opam monorepo` etc. -- but with MirageOS4, e.g. the `opam config var` calls to invoke `pkg-config` are no longer necessary. And I'd be keen on removing the `opam` binary as a dependency from a unikernel build. At least some versions of dune executed `opam --version`, but I think that has been revised as well.

Anyways, I gathered the parts of opam that are needed (or used to be needed?) in a shell script https://github.com/roburio/orb/blob/next/opam.sh (which works fine to build many unikernels. see the https://builds.robur.coop overview).

RyanGibb commented:
> It seems rather counter-intuitive to me when the `mirage` workflow requires `opam` in anyway.

With this workflow we actually only use opam to resolve dependency versions. Nix provides binary or source (for the monorepo workflow) derivations for all the dependencies.

RyanGibb commented:
> So, if it is desired to use multiple CPU cores while building ocaml-gmp, why not resort to the opam package processor (https://github.com/haesbaert/ocaml-processor) that provides a function `cpu_count` and is available for all OCaml versions and platforms ocaml-gmp is supposed to work with.

Correct me if I'm wrong, but this would require adding ocaml as a dependency to the project?

hannesm commented:
@RyanGibb yes, this will add "ocaml" as a dependency to ocaml-gmp. But turns out, it already depends on `dune` `ocaml` and `conf-m4`, so I hardly see any issue. From my understanding, the whole idea of this project is to deliver `gmp` as something built by `dune`.

RyanGibb commented:
> @RyanGibb yes, this will add "ocaml" as a dependency to ocaml-gmp. But turns out, it already depends on `dune` `ocaml` and `conf-m4`, so I hardly see any issue. From my understanding, the whole idea of this project is to deliver `gmp` as something built by `dune`.

Great, this sounds like it would be a more portable solution then :)

RyanGibb commented:
I can confirm this is building https://github.com/RyanGibb/mirage-www/!

dinosaure commented:
Can you run a `dune build @fmt --auto` on your PR? LGTM otherwise.

hannesm commented:
uhm, this still uses `/proc/cpuinfo` (thus, Linux-only) and conditionals...

RyanGibb commented:
> uhm, this still uses `/proc/cpuinfo` 

I've replaced this with a call to `ocaml-processor-dump` which we grep and cut.

> and conditionals...

And removed the opam conditional.

Let me know what you think of this!

hannesm commented:
thanks, looks good to me

RyanGibb commented:
Hmm, this doesn't seem to work in a monorepo:
```
www-dev> File "duniverse/ocaml-processor/lib/dune", line 1, characters 0-451:
www-dev>  1 | (library
www-dev>  2 |  (name processor)
www-dev>  3 |  (public_name processor)
www-dev> ....
www-dev> 12 |  (private_modules ioreg)
www-dev> 13 |  (foreign_stubs (language c) (names processor_stubs))
www-dev> 14 |  (c_library_flags (-lpthread)))
www-dev> Error: Some modules don't have an implementation.
www-dev> You need to add the following field to this stanza:
www-dev>   (modules_without_implementation query topology)
```

dinosaure commented:
/cc @haesbaert

haesbaert commented:
> Hmm, this doesn't seem to work in a monorepo:
> 
> ```
> www-dev> File "duniverse/ocaml-processor/lib/dune", line 1, characters 0-451:
> www-dev>  1 | (library
> www-dev>  2 |  (name processor)
> www-dev>  3 |  (public_name processor)
> www-dev> ....
> www-dev> 12 |  (private_modules ioreg)
> www-dev> 13 |  (foreign_stubs (language c) (names processor_stubs))
> www-dev> 14 |  (c_library_flags (-lpthread)))
> www-dev> Error: Some modules don't have an implementation.
> www-dev> You need to add the following field to this stanza:
> www-dev>   (modules_without_implementation query topology)
> ```
> 
> I could try and patch/submit a PR for this but I'm not familiar with the project. Do you know what's going on here (why these modules don't have an implementation) @hannesm?

Basically we try to do some discovery on how to best build the topology and query modules, somehow you're not falling into any of the cases from dune:
```
(rule
 (target query.ml)
 (enabled_if (= %{system} "macosx"))
 (action (copy query_apple.ml query.ml)))
(rule
 (target query.ml)
 (enabled_if (or
              (= %{system} "linux")
              (= %{system} "freebsd")
              (= %{system} "netbsd")
              (= %{system} "openbsd")))
 (action (copy query_from_topology.ml query.ml)))

(rule
 (target topology.ml)
 (enabled_if (and (= %{architecture} "amd64")
              (or (= %{system} "linux") (= %{system} "freebsd"))))
 (action (copy topology_amd64.ml topology.ml)))
(rule
 (target topology.ml)
 (enabled_if (and (= %{system} "macosx") (= %{architecture} "amd64")))
 (action (copy topology_apple_amd64.ml topology.ml)))
(rule
 (target topology.ml)
 (enabled_if (and (= %{system} "macosx") (= %{architecture} "arm64")))
 (action (copy topology_apple_arm64.ml topology.ml)))
(rule
 (target topology.ml)
 (enabled_if (or
              (= %{system} "netbsd")
              (= %{system} "openbsd")
              (= %{system} "dragonflybsd")
              (and (= %{system} "linux") (not (= %{architecture} "amd64")))))
 (action (copy topology_fake.ml topology.ml)))
```

Do the dune variables somehow evaluate to something else in the monorepo ? as in is system not linux and architecture not amd64 ?

hannesm commented:
May be related to / an instance of https://github.com/ocaml/dune/issues/3917

haesbaert commented:
> May be related to / an instance of [ocaml/dune#3917](https://github.com/ocaml/dune/issues/3917)

hmm could be, maybe @TheLortex can cheap in ? I never used monorepo, but to my knowledge those variables are obtained in the runtime (it's what comes from `ocamlc -config`)

RyanGibb commented:
> Basically we try to do some discovery on how to best build the topology and query modules

Thanks, that makes sense.

> Do the dune variables somehow evaluate to something else in the monorepo ? as in is system not linux and architecture not amd64 ?

Hmm, this is running on x86-64 Linux. Possibly nix sandboxing is to blame. I'm trying to figure out a way to debug dune variables (e.g. print them to stdout).

> (which nproc > /dev/null 2>&1 && nproc) || sysctl -n hw.ncpu

The only issue with this is that `nproc` prints the number of virtual cores rather than physical cores which may be sub-optimal for CPU bound processes (which I think building gmp is).

haesbaert commented:
> 
Not sure I agree, threads helping or not is not much related to it being CPU bound, it's more related to instructions per cycles and therefore cache-misses, building stuff is very cache non friendly, so threading should help.

This isn't something you/we should ponder much, this is hard to know in advance and depends on a bunch of stuff, the sane thing to do is just build on every logical CPU available (you can test if you want, but I doubt you find a case where threading doesn't help). 

I used to build many things on a Niagra T2 (8 threads per core) on an SO where the kernel was completely giant locked, and using all threads were still almost always beneficial.

RyanGibb commented:
That makes complete sense. It sounds like to avoid the complexity of building `ocaml-processor` from source with `opam monorepo` we should use your portable script to get the number of logical CPUs.

hannesm commented:
yes, well, at least on FreeBSD that's fine. :)

RyanGibb commented:
Great :) I've also added a default value of 1 if `sysctl` fails. Thanks for the help again @haesbaert !

haesbaert commented:
> Great :) I've also added a default value of 1 if `sysctl` fails, and squashed the commits together. Thanks for the help again @haesbaert !

Good call !
You're most welcome :dancing_men:

dinosaure commented:
Ok, so if everyone is happy with this PR, we can merge it. @TheLortex do you want to release it?

TheLortex commented:
Yes I can take care of it today.
